### ISSUE TITLE
Infinite recursion creating model for sub types with forward ref to data class

### ISSUE THREAD

**Issue by 0rd0g** [pydantic/pydantic#4792]
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://pydantic-docs.helpmanual.io) and still think this is a bug
- [X] I am confident that the issue is with pydantic (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

The snippet below shows a case of infinite recursion when the annotation for a field of a `dataclass` contains a forward ref to that same `dataclass`.

Note that if I change `Portfolio` instead to be a class derived from `BaseModel` it works ... but I think this is slightly by accident. The issue seems to be what is returned by `typing._eval_type()` for forward refs: if I made the `Portfolio` class below a `BaseModel` type then the pydantic model would be created before `Portfolio` had finished evaluating, and `typing._eval_type()` would continue to return the forward ref.

With the `dataclass` approach, `Portfolio` has already been created by the time I call `schema_json_of()` and `typing._eval_type()` returns the resolved `Portfolio` type.
<br/>

Please note that this is a simplified example from a large project and wholesale conversion from `dataclass` to `BaseModel` is not practical.

### Example Code

```Python
from dataclasses import dataclass
from pydantic import Field, schema_json_of
from typing import Annotated, ForwardRef, Literal, Tuple, Union


@dataclass
class FXInstrument:
    pair: str

    type_: Literal['FX'] = 'FX'


@dataclass
class RatesInstrument:
    ccy: str

    type_: Literal['Rates'] = 'Rates'


@dataclass
class EquityInstrument:
    underlier: str

    type_: Literal['Equity'] = 'Equity'


PriceableType = Annotated[Union[FXInstrument, RatesInstrument, EquityInstrument, ForwardRef('Portfolio')],
                          Field(discriminator='type_')]


@dataclass
class Portfolio:
    priceables: Tuple[PriceableType, ...]

    type_: Literal['Portfolio'] = 'Portfolio'


portfolio_schema = schema_json_of(Portfolio)  # Crashes - infinite recursion
```

### Highly Inadvisable Workaround

This hack will force it to 'work' by temporarily removing `Portfolio` from `globals`, forcing the pydantic model to build, then re-adding `Portfolio`. Naturally I would prefer a solution other than this ...

```Python
from pydantic import create_model,


__Portfolio = globals().pop(Portfolio.__name__)
m = create_model(__Portfolio.__name__, __root__=(__Portfolio, ...))
globals()[__Portfolio.__name__] = __Portfolio
m.update_forward_refs()
```

### Python, Pydantic & OS Version

```Text
pydantic version: 1.10.2
            pydantic compiled: True
                 install path: /Users/Nick/miniforge3/lib/python3.9/site-packages/pydantic
               python version: 3.9.1 | packaged by conda-forge | (default, Dec  9 2020, 01:07:47)  [Clang 11.0.0 ]
                     platform: macOS-12.4-arm64-arm-64bit
     optional deps. installed: ['typing-extensions']
```

### Affected Components

- [ ] [Compatibility between releases](https://pydantic-docs.helpmanual.io/changelog/)
- [ ] [Data validation/parsing](https://pydantic-docs.helpmanual.io/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://pydantic-docs.helpmanual.io/usage/exporting_models/) - `.dict()` and `.json()`
- [X] [JSON Schema](https://pydantic-docs.helpmanual.io/usage/schema/)
- [X] [Dataclasses](https://pydantic-docs.helpmanual.io/usage/dataclasses/)
- [ ] [Model Config](https://pydantic-docs.helpmanual.io/usage/model_config/)
- [ ] [Field Types](https://pydantic-docs.helpmanual.io/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://pydantic-docs.helpmanual.io/usage/validation_decorator/)
- [ ] [Generic Models](https://pydantic-docs.helpmanual.io/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://pydantic-docs.helpmanual.io/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://pydantic-docs.helpmanual.io/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by samuelcolvin** [comment]
This should not be a problem in V2 (once we implemented dataclasses, again see #4670) since recursive models are much more sophisticated in V2.

Again, I'd be willing to review a fix for this if it's very simple but I'm loath to did too deep into it in V1 for the risk of causing other changes in behaviour.

@PrettyWood would you agree?

**Comment by 0rd0g** [comment]
Understood. I may have a stab at fixing this and submitting a PR - has to be better than my globals hack!
